libdrgn/linux_kernel_helpers.c

```diff
@@ -533,3 +547,401 @@ struct drgn_error *linux_helper_find_task(struct drgn_object *res,
+	drgn_object_init(&iter->pos, prog);
+	drgn_object_init(&iter->ns, prog);
+	drgn_object_init(&iter->entry, prog);
+	err = drgn_program_find_object(prog, "pid_hash", NULL,
+				       DRGN_FIND_OBJECT_VARIABLE, &iter->pid_hash);
+	if (err)
+		goto out;
+	struct drgn_qualified_type void_star_type;
+	err = drgn_program_find_type(prog, "void *", NULL, &void_star_type);
+	if (err)
+		goto out;
+	err = drgn_object_set_unsigned(&iter->pos, void_star_type, 0, 0);
+	if (err)
+		goto out;
+	err = drgn_object_copy(&iter->ns, ns);
+	if (err)
+		goto out;
+	err = drgn_program_find_type(prog, "struct pid", NULL, &iter->pid_type);
+	if (err)
+		goto out;
+	err = drgn_program_find_type(prog, "struct upid", NULL,
+				     &iter->upid_type);
+	if (err)
+		goto out;
+	err = drgn_program_find_object(prog, "pidhash_shift", NULL,
+				       DRGN_FIND_OBJECT_ANY, &pidhash_shift);
+	if (err)
+		goto out;
+	err = drgn_object_read_unsigned(&pidhash_shift, &iter->index);
+	if (err)
+		goto out;
+	iter->index = 1 << iter->index;
+	err = drgn_object_member_dereference(&ns_level, ns, "level");
+	if (err)
+		goto out;
+	uint64_t ns_level_value;
+	err = drgn_object_read_unsigned(&ns_level, &ns_level_value);
+	if (err)
+		goto out;
+	snprintf(iter->member_specifier, sizeof(iter->member_specifier),
+		 "numbers[%" PRIu64 "]", ns_level_value);
+out:
+	if (err) {
+		drgn_object_deinit(&iter->pid_hash);
+		drgn_object_deinit(&iter->pos);
+		drgn_object_deinit(&iter->ns);
+		drgn_object_deinit(&iter->entry);
+	}
+	drgn_object_deinit(&pidhash_shift);
+	drgn_object_deinit(&ns_level);
+	return err;
+}
+
+struct drgn_error *pid_iter_init(struct pid_iter *iter,
+				 const struct drgn_object *ns)
+{
+	struct drgn_program *prog = drgn_object_program(ns);
+	struct drgn_error *err;
+	struct drgn_object idr;
+	drgn_object_init(&idr, prog);
+
+	err = drgn_object_member_dereference(&idr, ns, "idr");
+	if (!err) {
+		iter->has_idr = true;
+		err = drgn_program_find_type(prog, "struct pid *", NULL,
+					     &iter->pid_type);
+		if (!err)
+			err = idr_iter_init(&iter->iter, &idr);
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		drgn_error_destroy(err);
+		err = pid_iter_init_old_kernel(prog, ns, iter);
+	}
+
+	drgn_object_deinit(&idr);
+	return err;
+}
+
+void pid_iter_deinit(struct pid_iter *iter)
+{
+	if (iter->has_idr) {
+		idr_iter_deinit(&iter->iter);
+	} else {
+		drgn_object_deinit(&iter->pid_hash);
+		drgn_object_deinit(&iter->pos);
+		drgn_object_deinit(&iter->ns);
+		drgn_object_deinit(&iter->entry);
+	}
+}
+
+struct drgn_error *pid_iter_next(struct pid_iter *iter,
+				 struct drgn_object **ret)
+{
+	if (iter->has_idr) {
+		struct radix_tree_iter_entry *entry;
+		struct drgn_error *err = idr_iter_next(&iter->iter, &entry);
+		if (err)
+			return err;
+		if (!entry) {
+			*ret = NULL;
+			return NULL;
+		}
+		err = drgn_object_cast(&entry->node, iter->pid_type,
+				       &entry->node);
+		*ret = err ? NULL : &entry->node;
+		return err;
```

Comment:
Sure, was trying to be defensive, didn't realize there was an established convention on _all_ out params not being modified on error.

Reply:
It's more that we generally assume that if a function returns an error, the contents of its out parameters are undefined unless it's documented otherwise. `drgn_object` out parameters are a documented exception. So, other than `drgn_object`:

* A function is free to do whatever it wants to the out parameters in the middle of the function as long as on success it "does what it's supposed to do".
* A caller should not access the out parameters if the function it called failed.